homeassistant/components/simplisafe/camera.py

```diff
@@ -0,0 +1,233 @@
+from simplipy.errors import SimplipyError
+from simplipy.system.v3 import SystemV3
+from simplipy.websocket import EVENT_CAMERA_MOTION_DETECTED, WebsocketEvent
+import voluptuous as vol
+
+from homeassistant.components.camera import Camera
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_ENTITY_ID
+from homeassistant.core import HomeAssistant, ServiceCall, callback
+from homeassistant.exceptions import HomeAssistantError
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.template import Template
+
+from . import SimpliSafe, SimpliSafeEntity
+from .const import DOMAIN, LOGGER
+
+DEFAULT_IMAGE_WIDTH = 720
+
+# The Outdoor Camera has a custom service for saving the motion clip.
+ATTR_FILENAME = "filename"
+ATTR_WIDTH = "width"
+
+SERVICE_OC_IMAGE = "capture_motion_image"
+SERVICE_OC_CLIP = "capture_motion_clip"
+
+SERVICE_OC_IMAGE_SCHEMA = cv.make_entity_service_schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_ids,
+        vol.Required(ATTR_FILENAME): cv.template,
+        vol.Optional(ATTR_WIDTH, default=720): vol.Coerce(int),
+    }
+)
+
+SERVICE_OC_CLIP_SCHEMA = cv.make_entity_service_schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_ids,
+        vol.Required(ATTR_FILENAME): cv.template,
+    }
+)
+
+SERVICES = (
+    SERVICE_OC_IMAGE,
+    SERVICE_OC_CLIP,
+)
+
+
+def _write_image(to_file: str, image_data: bytes | None) -> None:
+    """Write image content to a file."""
+    if image_data is None:
+        return
+    os.makedirs(os.path.dirname(to_file), exist_ok=True)
+    with open(to_file, "wb") as img_file:
+        img_file.write(image_data)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up SimpliSafe cameras based on a config entry."""
+    simplisafe = hass.data[DOMAIN][entry.entry_id]
+
+    cameras: list[SimplisafeOutdoorCamera] = []
+
+    for system in simplisafe.systems.values():
+        if system.version == 2:
+            LOGGER.info("Skipping camera setup for V2 system: %s", system.system_id)
+            continue
+
+        for camera in system.cameras.values():
+            if camera.camera_type == CameraTypes.OUTDOOR_CAMERA:
+                outdoor = SimplisafeOutdoorCamera(hass, simplisafe, system, camera)
+                for service, method, schema in (
+                    (
+                        SERVICE_OC_IMAGE,
+                        outdoor.save_image_handler,
+                        SERVICE_OC_IMAGE_SCHEMA,
+                    ),
+                    (
+                        SERVICE_OC_CLIP,
+                        outdoor.save_clip_handler,
+                        SERVICE_OC_CLIP_SCHEMA,
+                    ),
+                ):
+                    hass.services.async_register(DOMAIN, service, method, schema=schema)
```

Comment:
We're registering a service for every camera?

Reply:
My service handlers need a "self" that points to a specific camera.

Reply:
`self` will be provided by the entity platform. Take a look at how RainMachine does it. Here's where the services are registered:

https://github.com/home-assistant/core/blob/e3680044fedc253ec00649103fad5fe34ae4f72a/homeassistant/components/rainmachine/switch.py#L182-L196

...and here's an example where the handlers are registered:

https://github.com/home-assistant/core/blob/e3680044fedc253ec00649103fad5fe34ae4f72a/homeassistant/components/rainmachine/switch.py#L348-L357

Crucially, note that I don't register the service for every entity.

Reply:
I believe I am following that pattern exactly, but my handlers are not being called.